### PR TITLE
Add ICL datasets, plus functionality for subcategories in datasets

### PR DESCRIPTION
**Pull request by abhi-mosaic**
This is basically a copy of #9 but merged with latest folder restructure, and passing tests and linting.

**from previous PR:**

This diff updates us to use all of the new composer ICL features: schema datasets, QA tasks, and sub categories.

It checks in standardized versions of our datasets and standardizes the configs we use to run eval.

TO DO: I am going to add some unit testing for the dataset formatting under a variety of tokenizers (GPT NeoX, LLaMa, OPT, GPT2)

### CODE REVIEW

**Review by bmosaicml — COMMENTED** [review]
LGTM...do you want me to abandon my PR and let you take over?

**Review by bmosaicml — APPROVED** [review]

File: llmfoundry/utils/builders.py [hunk 1]
```diff
@@ -170,6 +178,12 @@ def _validate_cfg(icl_cfg):
                 icl_cfg.metric_names = [
                     'InContextLearningMultipleChoiceAccuracy'
                 ]
+            elif icl_cfg.icl_task_type == 'schema':
+                icl_cfg.metric_names = [
+                    'InContextLearningMultipleChoiceAccuracy'
+                ]
+            elif icl_cfg.icl_task_type == 'question_answering':
+                icl_cfg.metric_names = ['InContextLearningQAAccuracy']
```
Comment (abhi-mosaic) [inline llmfoundry/utils/builders.py:186]: @bmosaicml confirming this is ok?

File: scripts/eval/yamls/tasks.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
+icl_tasks:
+-
+  label: jeopardy
+  dataset_uri: eval/local_data/jeopardy_all.jsonl # ADD YOUR OWN DATASET URI
+  num_fewshot: [0, 1, 5, 10]
+  icl_task_type: language_modeling
+  continuation_delimiter: "\nAnswer: " # this separates questions from answers
```
Comment (abhi-mosaic) [inline scripts/eval/yamls/tasks.yaml:7]: @bmosaicml could you spot check all the continuation delimiter choices here? The default is "\n" but then some of these tasks seem to use different ones like `Answer: ` or `\nAnswer: `

File: scripts/eval/README.md [hunk 3]
```diff
@@ -1,9 +1,9 @@
-# LLM Evaluation
-This is the MosaicML LLM evaluation suite. It is the world's fastest, multi-GPU compatible in-context-learning (ICL) evaluaton suite with support for using [FSDP](https://engineering.fb.com/2021/07/15/open-source/fsdp/) with any model on the HuggingFace hub or for any PyTorch model that implements the MPT model interface.
+# In-context learning (ICL) evaluaton
+This is the MosaicML LLM evaluation suite. It is the world's fastest, multi-GPU compatible ICL evaluaton suite with built in support for using [FSDP](https://engineering.fb.com/2021/07/15/open-source/fsdp/) with any model on the HuggingFace hub or for any PyTorch model that implements the MosaicGPT model interface.
```
Comment (dakinggg) [inline scripts/eval/README.md:2]: FSDP link should maybe be a link to the pytorch docs at this point?

Comment (dakinggg) [inline scripts/eval/README.md:2]: what "model interface" is this supposed to say? maybe `Composer`?

File: scripts/eval/README.md [hunk 4]
```diff
@@ -74,6 +75,7 @@ Below is a complete YAML section that works with the TriviaQA dataset in `llmfou
     example_delimiter: '\n' # this goes between fewshot examples
```
Comment (dakinggg) [inline scripts/eval/README.md:75]: all the new lines here should change to using double quotes

File: scripts/eval/README.md [hunk 5]
```diff
@@ -204,12 +211,14 @@ Below is a YAML section that works with the Winograd dataset in `llmfoundry/icl_
...
     continuation_delimiter: ' ' # this separates questions from answers
 >
+
 ### Build your own dataset (BYOD)
 Building a dataset compatible with our eval suite is very easy if it fits with one of the four supported task types. Simply choose the appropriate task type (LM, MC, QA, or Schema) and process each dataset into a jsonl format where each row has the format described above.
 
 Below is a minimal script which prepares the [Winograd schema challenge](https://cdn.aaai.org/ocs/4492/4492-21843-1-PB.pdf) hosted on [HuggingFace](https://huggingface.co/datasets/winograd_wsc). This script can be modified to generate other datasets based on the HuggingFace dataset hub.
 ```jsx
```
Comment (dakinggg) [inline scripts/eval/README.md:219]: syntax here for the code block should be python

File: scripts/eval/yamls/mpt_eval.yaml [hunk 6]
```diff
@@ -1,5 +1,6 @@
 max_seq_len: 2048
 tokenizer_name: gpt2
```
Comment (dakinggg) [inline scripts/eval/yamls/mpt_eval.yaml:2]: tokenizer is wrong?

File: scripts/eval/yamls/tasks.yaml (hunk 2 above)
Comment (dakinggg) [reply scripts/eval/yamls/tasks.yaml:7]: i believe the default continuation delimiter is space, not sure how `Answer:` vs `\nAnswer` was decided

File: mcli/mcli-hf-eval.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
+integrations:
+- integration_type: git_repo
+  git_repo: mosaicml/llm-foundry
+  git_branch: icl_eval_categories # use your branch
```
Comment (dakinggg) [inline mcli/mcli-hf-eval.yaml:4]: update before merging probably

File: llmfoundry/utils/builders.py (hunk 1 above)
Comment (bmosaicml) [reply llmfoundry/utils/builders.py:186]: yep!

File: scripts/eval/yamls/mpt_eval.yaml (hunk 6 above)
Comment (bmosaicml) [reply scripts/eval/yamls/mpt_eval.yaml:2]: good point

**Comment by abhi-mosaic** [comment]
Going to fix all but the README and merge, flaky dataloader tests will be fixed by @mvpatel2000 in followup PR

File: scripts/eval/yamls/tasks.yaml (hunk 2 above)
Comment (bmosaicml) [reply scripts/eval/yamls/tasks.yaml:7]: I just double checked them all they're looking correct. Some datasets have newlines appended to the query so there's no need to include it in the Answer delimiter

**Review by bmosaicml — APPROVED** [review]

**Pull request merged**